module added for schedule

nirupaman opened the pull request:
root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# ansible-playbook schedule.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

 [WARNING]: While constructing a mapping from /home/nirun/niru_ansiblecopy/hpe3par_ansible_module/remote_copy.yml, line 3,
column 5, found a duplicate dict key (http_proxy). Using last defined value only.


PLAY [localhost] ************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [Create Schedule on source] ********************************************************************************************
changed: [localhost]

TASK [pause] ****************************************************************************************************************
[pause]
Press Enter:
ok: [localhost]

TASK [Delete Schedule from source] ******************************************************************************************
changed: [localhost]

PLAY RECAP ******************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0

nirupaman commented:
@farhan7500 Please review

farhan7500 commented:
Can you also paste the CLI output showing the schedule has been created

farhan7500 commented on Modules/hpe3par_snapshot.py:
```diff
@@ -435,13 +447,131 @@ def restore_snapshot_online(
+        client_obj.login(storage_system_username, storage_system_password)
+        client_obj.setSSHOptions(storage_system_ip, storage_system_username, storage_system_password)
+        expirationHours = convert_to_hours(expiration_time, expiration_unit)
+	retentionHours = convert_to_hours(retention_time, retention_unit)
```
Is this indented correctly?

farhan7500 commented on Modules/hpe3par_snapshot.py:
```diff
@@ -435,13 +447,131 @@ def restore_snapshot_online(
+        expirationHours = convert_to_hours(expiration_time, expiration_unit)
+	retentionHours = convert_to_hours(retention_time, retention_unit)
+        if client_obj.volumeExists(base_volume_name):
+           cmd = ["\"createsv"] 
```
What happens when the schedule that we are creating already exists? We need to check if a schedule is created by that name. Therefore, the following methods need to be added in SDK and client

a. getSchedule(name)
b. scheduleExists(name)

scheduleExists will refer try and get the schedule and based on that result we have to decide in ansible whether to create the new schedule

nirupaman replied:
If schedule exist it show error that schedule already exist ..Can we have only one function scheduleExists(name) in sdk

farhan7500 commented on Modules/hpe3par_snapshot.py:
```diff
@@ -435,13 +447,131 @@ def restore_snapshot_online(
+        client_obj.login(storage_system_username, storage_system_password)
+        client_obj.setSSHOptions(storage_system_ip, storage_system_username, storage_system_password)
+        if client_obj.scheduleExists(schedule_name):
+           client_obj.deleteSchedule(schedule_name)
```
Same again. What happens if the schedule does not exist and we try to delete it

nirupaman replied:
handled

nirupaman commented on Modules/hpe3par_snapshot.py:
```diff
@@ -117,11 +117,21 @@
+    required: true
+  task_freq:
+    description:
+      - "Frequency of the schedule to be created."
```
added option for task_freq

nirupaman commented:
Success Output

1) pytest run
root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# pytest test/unit/test_hpe3par_snapshot.py
=====================  test_hpe3par_snapshot.py  =====================
...............
----------------------------------------------------------------------
Ran 15 tests in 0.025s

OK
*******************************************************************************

2) Playbook create and delete schedule

Create Schedule

root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# ansible-playbook playbooks/create_sch_playbook.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

 [WARNING]: While constructing a mapping from
/home/nirun/niru_ansiblecopy/hpe3par_ansible_module/playbooks/create_sch_playbook.yml, line 3, column 5, found a duplicate
dict key (http_proxy). Using last defined value only.


PLAY [localhost] ************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [Create Schedule on source] ********************************************************************************************
changed: [localhost]

PLAY RECAP ******************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module#
==============================================================
Delete Schedule

root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# ansible-playbook playbooks/delete_sch_playbook.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

 [WARNING]: While constructing a mapping from
/home/nirun/niru_ansiblecopy/hpe3par_ansible_module/playbooks/delete_sch_playbook.yml, line 3, column 5, found a duplicate
dict key (http_proxy). Using last defined value only.


PLAY [localhost] ************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [Delete Schedule from source] ******************************************************************************************
changed: [localhost]

PLAY RECAP ******************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

farhan7500 commented on Modules/hpe3par_snapshot.py:
```diff
@@ -505,7 +637,18 @@ def main():
+        },
+        "task_freq": {
+            "type": "str",
+            "choices": ['yearl', 'monthly', 'weekly', 'daily', 'hourly']
```
This does not take into account cron type option

And I guess it must be yearl'y'

nirupaman replied:
changed

nirupaman commented on Modules/hpe3par_snapshot.py:
```diff
@@ -435,13 +447,134 @@ def restore_snapshot_online(
+        client_obj.setSSHOptions(storage_system_ip, storage_system_username, storage_system_password)
+        expirationHours = convert_to_hours(expiration_time, expiration_unit)
+	retentionHours = convert_to_hours(retention_time, retention_unit)
+        if client_obj.volumeExists(base_volume_name):
```
We also need to check if volume exists

nirupaman commented on Modules/hpe3par_snapshot.py:
```diff
@@ -435,13 +447,134 @@ def restore_snapshot_online(
+        if client_obj.volumeExists(base_volume_name):
+           cmd = ["createsv"] 
+	   if read_only:
+              cmd.append("-ro")
```
which value?

nirupaman commented:
@farhan7500 : Please review added scheduleExists and try except in getSchedule

farhan7500 commented:
Please remove ALL unrelated changes from the PR. Only then will I be able to review properly

nirupaman commented:
Duplicate
